### ISSUE TITLE
[Feature] Historical Backfilling and Rebasing of Snapshots

### ISSUE THREAD

**Issue by walker-philips** [dbt-labs/dbt-core#9892]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

My understanding: Snapshots are intended to record historical data safely, and avoid accidental deletion of data. They are driven by timestamps or data driven deltas specified by the user. Data driven Snapshots cannot "replay" history, such that a Snapshot query should only return a single row to represent a desired unit of data. As an example, a Snapshot Check Strategy query of Bank Account balances should only show the current day's balance per account as opposed to several days worth of balances per  account.

Users of DBT may possess lower/same quality of historical prior to a Snapshot being formally initialized. Currently, there is no way to backfill a Snapshot. Also, a user may change their approach to Snapshotting data, causing mass, unnecessary snapshotting of rows or business rules may change leading to reducing the depth of snapshots. There is no way to "condense" snapshots by removing rows of now-duplicate data and updating the dbt_valid_from/dbt_valid_to columns to remain linear.

### Describe alternatives you've considered

I have a rough working approach utilizing a macro, a template model, a controlling model that coordinates the backfill of several Snapshots, pre/post hooks to run the necessary update statements. It does not attempt to alter the existing Snapshot. Instead, you could feed the existing Snapshot as a backfill to a new Snapshot model. In order to promote data survivability, a user can choose to save backfill data as a table to allow recovery.

### Who will this benefit?

Any users with historical data they would like to merge into a Snapshot or one who wants to rebase/condense an existing Snapshot due to a change in Snapshot strategy/methodology

### Are you interested in contributing this feature?

Yes, I have an example of my imperfect approach

**Comment by walker-philips** [comment]
@dbeatty10 Let me know if this is sufficient for lining up the feature? Happy to provide my scripts as a point of reference.

**Comment by dbeatty10** [comment]
Thanks for opening this @walker-philips 🤩

Yes, any additional points of reference (such as your scripts) would be great to see as well.

**Comment by walker-philips** [comment]
@dbeatty10 Where would be the best place to add those? I didnt want to fork just for a few files and I also cannot add my full project due to compliance reasons.

**Comment by dbeatty10** [comment]
Here's a few ideas:

- Share the files here (optionally wrapping in a [Markdown toggle](https://gist.github.com/Ashton-W/864d09d71f7cb8e5b47ebaac02b98455) for readability)
- Share the files in a [GitHub Gist](https://gist.github.com/)
- [Fork in GitHub](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo) and make a commit

**Comment by jemimamace** [comment]
Voicing my support for this one, recently had a similar use case for snapshots needing to be 'backfilled'.

**Comment by dbeatty10** [comment]
@walker-philips we're waiting on additional points of reference (such as your scripts) to assess this feature request further. Are you still planning on sharing those?

We've got an initiative right now related to snapshots, so it would be good timing for us to consider your ideas if you are able to provide more detail. You can read more here:

- [GitHub epic](https://github.com/dbt-labs/dbt-core/issues/10151)
- [GitHub Discussion](https://github.com/dbt-labs/dbt-core/discussions/7018)

Without additional information of what this feature request would look like, this is issue is likely to go stale again.

**Comment by walker-philips** [comment]
@dbeatty10

So here's my assumptions and approach:
An existing snapshot implies there is:
- a key of some kind
- a dbt_valid_from/timestamp value that creates some level of order to the data
- some rules regarding what will trigger a snapshot ( a column(s) updating for a timestamp strategy or a subset of columns for a check strategy) 
- a place(s) where the snapshot originated from
- a current structure of the snapshot (table schema/column types)

My assumption:
- The end user is going to have to transform the historical/deprecating snapshot data to some degree to adhere to the current/desired structure of a snapshot
- The user may combine the existing snapshot and as many other data sources together to create a historical dataset
- The user can use existing dbt_valid_from values and guesstimate older data's validity start times
- The user will prefer to avoid oversnapshotting data so they can use group by's to slim the data down to rows that accurately represent the level of snapshotting desired
- Valid to is implicitly tied to the next unique_key's row's valid from value.
- People would like to avoid the costly efforts of backfilling each time they ```dbt build``` so we should save the transformed backfill dataset as an audit trail/and optimization for speed over storage.
- dbt model names are unique and thus we can use hacky dynamic model naming to handle creating these supplementary models

The approach works as follows:
the render_backfills model takes a list of the existing snapshot model names along with their unique key(s) as dictionary. It will iterate through them to coordinate the creation and loading of backfills by calling the apply_snapshot_backfill() macro. This will load data into the existing snapshot table from the newly created backfill data source.

For each snapshot to be backfilled, we have to create a backfill model that transforms all the historical data into a form that can be unioned with the current snapshot table. It has some necessary criteria that always has to be met. Namely the designation of a unique_key and dbt_valid_from stand in.

I utilize pre/post hooks to execute INSERT statements that save the backfill data as its own table disconnected from the dbt framework. This helps avoid circular dependencies.

The dbt_backfill_applier_model() is simply boilerplate code that helps determine whether to reuse an instantiated backfill dataset or make it for the first time.

Lastly, render_backfills deletes itself as its essentially useless from a database asset standpoint. It has to be a table in order to force the triggering of the post hook and execution of the run_query command


Admittedly, this feels super hacky. But they key takeaways I have are:
1. The transformation of historical datasets into snapshot merge-worthy models seem to be a great usage of DBT.
2. Slimming/expanding datasets to new snapshot rules doesn't seem inherently complex.
3. There should be no limitation on what datasets an end user may choose to merge into a snapshot. They do so at their own risk and should have proper backups to recreate their snapshots if all goes wrong.
4. We should avoid what could be a costly operation to assemble these backfill datasets
5. The complexity of setting up my approach has much to be improved
6. It should also be possible to do snapshot pruning by feeding the current snapshot into itself, however I think the inherent circular dependency may require hardcoding model names rather than using ref()

[apply_snapshot_backfill.txt](https://github.com/user-attachments/files/16432310/apply_snapshot_backfill.txt)
[dbt_backfill_applier_model.txt](https://github.com/user-attachments/files/16432311/dbt_backfill_applier_model.txt)
[my_table_name_backfill_apply.txt](https://github.com/user-attachments/files/16432312/my_table_name_backfill_apply.txt)
[render_backfills.txt](https://github.com/user-attachments/files/16432313/render_backfills.txt)
[column_intersect.txt](https://github.com/user-attachments/files/16432314/column_intersect.txt)
[my_table_name.txt](https://github.com/user-attachments/files/16432315/my_table_name.txt)

**Comment by walker-philips** [comment]
@dbeatty10 

Realized I was missing something like this to handle consecutive, repeating data. There's a potential that by just grouping by the check_cols (for a check strategy at least), you would eliminate future, repeated values entirely.

```
{% macro identify_consecutive_duplicates_cte(source_cte_name, check_cols, pk, date_col) %}
select *,
	(case 
		when LAG(concat_ws('-',{{check_cols|join('::text,')}}::text)) over (partition by {{ pk }} ORDER BY {{ date_col }} asc) 
			= concat_ws('-',{{check_cols|join('::text,')}}::text)	
		then true
		else false
	end) as is_duplicate
from {{source_cte_name}}
{%- endmacro %}
```

**Comment by fbardos** [comment]
Also raising my voice for this kind of feature. On my day job, we have a lot of exports once made from the database.
These exports are just plain exports from the database (full table, no `updated_at` column), so I use DBT to evaluate which rows have changed between exports.
I want to feed them back into DBT to use the potential of a SCD-Type-2 table, mainly to not store the same information multiple times. Additionally, I would like to be able to "regenerate" the whole snapshot table when something went wrong, based from these once made exports.

So, in summary, I want to use the following snapshot features together:
* strategy: check
* invalidate_hard_deletes: True
* a custom timestamp in the past which will then be used to set `dbt_valid_from`, `dbt_valid_to` and `dbt_updated_at`

Out of the box, this is not possible as far as I know, because for example, when calculating the deletes in the macro `snapshot_staging_table()`,  `dbt_valid_from`, `dbt_valid_to` and `dbt_updated_at` will default to `snapshot_get_time()`, which is the current time, not the time of the past export.

To achieve the desired behaviour, I have overwritten the macro `snapshot_get_time()` as follows:

```sql
{%- macro snapshot_get_time() -%}
    {% if var("snapshot_custom_date", false)  %}
        {{ print('--- OVERWRITE MACRO snapshot_get_time() ---') }}
        '{{ var("snapshot_custom_date") }}'::date
    {% else %}
        {{ adapter.dispatch('snapshot_get_time', 'dbt')() }}
    {% endif %}
{%- endmacro -%}
```

In my orchestrator (Dagster), I then feed the initialization of `dbt snapshot` with custom `--vars`:

```python
yield from dbt_cli.cli(
    [
        "snapshot",
        "--select", "my_snapshot_containing_backfills",
        "--vars", f"snapshot_custom_date: {_timestamp}"
    ],
    context=context
).stream()
```

I wrote a small test with 3 imaginary exports from the past with the following timestamps:
* 2025-01-01
* 2025-02-01
* 2025-03-01

Creating the snapshot with:

```sql
{% snapshot test_custom_snapshot_snapshot %}
{{
    config(
        target_database='mydb',
        target_schema='snapshots',
        unique_key='subjekt_id',
        strategy='check',
        check_cols=[
            'stadtkreis_bfs_id',
        ],
        invalidate_hard_deletes=True
    )
}}
select
    subjekt_id
    , stadtkreis_bfs_id
    , stadtkreis
    , use_case
from {{ source('qa', 'test_custom_snapshot') }}
{% endsnapshot %}
```

The snapshot table seems to work now, looking like this:

```text
mydb=# select subjekt_id, stadtkreis_bfs_id, stadtkreis, use_case, dbt_updated_at, dbt_valid_from, dbt_valid_to from snapshots.test_custom_snapshot_snapshot order by subjekt_id, dbt_valid_from;
 subjekt_id | stadtkreis_bfs_id | stadtkreis |                                 use_case                                 | dbt_updated_at | dbt_valid_from | dbt_valid_to
------------+-------------------+------------+--------------------------------------------------------------------------+----------------+----------------+--------------
          1 |                 1 | Stadt      | does never change                                                        | 2025-01-01     | 2025-01-01     |
          2 |                 1 | Stadt      | does disappear in second (2025-02-01)                                    | 2025-01-01     | 2025-01-01     | 2025-02-01
          3 |                 1 | Stadt      | does disappear in second (2025-02-01) and reappear in third (2025-03-01) | 2025-01-01     | 2025-01-01     | 2025-02-01
          3 |                 1 | Stadt      | does disappear in second (2025-02-01) and reappear in third (2025-03-01) | 2025-03-01     | 2025-03-01     |
          4 |                 1 | Stadt      | does change in second (2025-02-01)                                       | 2025-01-01     | 2025-01-01     | 2025-02-01
          4 |                 3 | Seen       | does change in second (2025-02-01)                                       | 2025-02-01     | 2025-02-01     |
          5 |                 1 | Stadt      | does change in second (2025-02-01) and change back in third (2025-03-01) | 2025-01-01     | 2025-01-01     | 2025-02-01
          5 |                 3 | Seen       | does change in second (2025-02-01) and change back in third (2025-03-01) | 2025-02-01     | 2025-02-01     | 2025-03-01
          5 |                 1 | Stadt      | does change in second (2025-02-01) and change back in third (2025-03-01) | 2025-03-01     | 2025-03-01     |
          6 |                 1 | Stadt      | does appear on second (2025-02-01) and never change                      | 2025-02-01     | 2025-02-01     |
(10 Zeilen)

```

Let me know if you found a more elegant solution, thank you. :)
